application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java

```diff
@@ -547,4 +555,63 @@ private void updateOAuth2Params() {
         log.warn("CAUTION: Update of Oauth2 parameters from 3.2.2 to 3.3.0 available only in ThingsBoard versions 3.3.0/3.3.1");
     }
 
+    private final PaginatedUpdater<String, Tenant> tenantCustomersTitleUpdater =
+            new PaginatedUpdater<>() {
+
+                @Override
+                protected String getName() {
+                    return "Customer title updater";
+                }
+
+                @Override
+                protected PageData<Tenant> findEntities(String id, PageLink pageLink) {
+                    return tenantService.findTenants(pageLink);
+                }
+
+                @Override
+                protected void updateEntity(Tenant entity) {
+                    updateDuplicateCustomersTitle(entity);
+                }
+            };
+
+    private List<Customer> updateDuplicateCustomersTitle(Tenant entity) {
+        customerDao.find(entity.getId());
+        List<Customer> customers = customerDao.find(entity.getId());
+        return updateDuplicateCustomersTitle(customers);
+    }
+
+    protected List<Customer> updateDuplicateCustomersTitle(List<Customer> customers) {
+        if (customers == null || customers.isEmpty()) return customers;
+        sortCustomersByTenantIdAndTitleAndCreatedTime(customers);
+        int countEqualsTitleAndTenantIdBefore = 0;
+        String lastCustomerName = customers.get(0).getName();
+        TenantId lastTenantId = customers.get(0).getTenantId();
+        for (int i = 1; i < customers.size(); i++) {
+            Customer customer = customers.get(i);
+            if (customer.getName().equals(lastCustomerName) && customer.getTenantId().equals(lastTenantId)) {
+                countEqualsTitleAndTenantIdBefore++;
+                customer.setTitle(customer.getName() + "-" + countEqualsTitleAndTenantIdBefore);
+                customers.set(i, updateCustomerTitle(customer.getTenantId(), customer));
+            } else {
+                countEqualsTitleAndTenantIdBefore = 0;
+            }
+            lastCustomerName = customer.getName();
+            lastTenantId = customer.getTenantId();
+        }
+        return customers;
+    }
+
+    protected Customer updateCustomerTitle(TenantId id, Customer customer) {
+        customerDao.removeById(id, customer.getUuidId());
+        return customerDao.save(id, customer);
+    }
+
+    protected void sortCustomersByTenantIdAndTitleAndCreatedTime(List<Customer> customers) {
+        customers.sort((o1, o2) -> {
+            if (!o1.getTenantId().equals(o2.getTenantId())) return o1.getTenantId().toString().compareTo(o2.getTenantId().toString());
```

Comment:
do you need to do this if all customers are related to single tenant here?

Reply:
Yes, I'm testing, and this code returns entities with a different tenantId

Reply:
Sorry, my mistake. Thanks for advice!